### DISCUSSION TITLE
Custom Redis Adapter support for session persistence

### DISCUSSION THREAD

**Discussion by sumodhnair19** [nextauthjs/next-auth#4987] [Help]
### Question 💬

Hi,

Thank you for such a fantastic library.  I was going through the documentation but didn't find any mention of connecting to custom configured `Redis` without the use of `Upstash` for a persistent `session store`. 

My use case is very simple. I am using `Nginx` as a load balancer between multiple nodes for my `nextJS` application and I would like to persist the `session` if in case the user logs in and refreshes the page as `Nginx` switches between nodes. I am using the credentials provider of `next-auth` as I have a Django-based auth system already available. 

I did see the implementation of the `next-auth `adapter with `Upstash`. However, I have my own custom server running with Redis.
 
I tried connecting to `redis` using `ioredis` which seems to connect. However, I am not sure how can I use `redis` here with `next-auth` to persist session and validate at the same time?

Thanks for the help.

### How to reproduce ☕️

```
import CredentialsProvider from "next-auth/providers/credentials";
import {UpstashRedisAdapter} from "@next-auth/upstash-redis-adapter";
import Redis from 'ioredis';

const redis = new Redis(process.env.REDIS_URL);

export const authOptions = {
    providers: [CredentialsProvider({
            name: 'auth',
            credentials: {
                email: {
                    label: 'email',
                    type: 'text',
                    placeholder: 'jsmith@example.com'
                },
                password: {
                    label: 'Password',
                    type: 'password'
                }
            },
            async authorize(credentials, req) {
                const payload = {
                    email: credentials.email,
                    password: credentials.password
                };

                const res = await fetch(`my-auth-system-url`, {
                    method: 'POST',
                    body: JSON.stringify(payload),
                    headers: {
                        'Content-Type': 'application/json'
                    }
                });

                const user = await res.json();
                console.log("user", user);
                if (!res.ok) {
                    throw new Error(user.exception);
                }
                // If no error and we have user data, return it
                if (res.ok && user) {
                    return user;
                }

                // Return null if user data could not be retrieved
                return null;
            }
        })],
    adapter: UpstashRedisAdapter(redis),
    pages: {
        signIn: '/login'
    },
    jwt: {
        secret: process.env.SECRET,
        encryption: true
    },
    callbacks: {
        jwt: async({token, user}) => {
            user && (token.user = user)
            return token
        },
        session: async({session, token}) => {
            session.user = token.user
            return session
        },
        async redirect({baseUrl}) {
            return `${baseUrl}/`
        }
    },
    session: {
        strategy: "jwt",
        maxAge: 3000
    },
    secret: process.env.SECRET,
    debug: true
}

export default NextAuth(authOptions)

```

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by galer7** [comment]
Link to the docs: https://next-auth.js.org/providers/credentials. It states that for `Credentials` provider, only JWT sessions are supported.
I'm also trying to find a way to persist JWT sessions from credentials login in the database, as I want to create some sort of online user search bar, and the only way to do that would be with a populated `Session` table.

I guess I'll try to implement a `Google` provider and ditch the `Credentials` one, but it would be cool to have a guide on self-maintained sessions in the database e.g. how to add, delete on browser tab close or sign out.

**Reply by sumodhnair19** [reply]
Thanks @galer7 . Yeah. It's unfortunate that such a great library doesn't provide support for `DB` for `session` persistence using `credentials`. I can't use google provider as we already have a custom authentication already built and have to rely on the `credentials provider`. Can you maybe suggest any other library I could use in tandem with `next-auth` to have a `database` for `session` persistence?
